### DISCUSSION TITLE
第4节课笔记(7班)

### DISCUSSION THREAD

**Discussion by JimmyMa99** [InternLM/Tutorial#310] [笔记]
第4节课笔记(7班)

**Comment by wuhaostudio** [comment]
https://blog.csdn.net/2301_77286822/article/details/135545724?spm=1001.2014.3001.5501

**Comment by createup2023** [comment]
https://www.yuque.com/creatup/ynrsde/wsrt4sgqo09epw54?singleDoc# 《第四次笔记》
补充CSDN笔记：https://blog.csdn.net/weixin_48339493/article/details/135756461

**Comment by leochiang** [comment]
第四次作业  https://vvgyb242x83.feishu.cn/wiki/EfeSwxynoiWNSWkU4ydcRqW6nkd?from=from_copylink

**Comment by hui1feng** [comment]
第四节课笔记：https://github.com/hui1feng/XTuner-Finetune/blob/main/README.md

**Comment by SaaRaaS-1300** [comment]
第四节课 笔记同作业 github链接：https://github.com/SaaRaaS-1300/InternLM_openNotebook/blob/main/Lesson-4/Lesson-4-Notebook.md

**Comment by PommesPeter** [comment]
# 第四节课笔记

> 博客链接：[https://memo.sylin.host](https://memo.sylin.host/docs/DeepLearning/%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86/%E8%AF%BE%E7%A8%8B%E5%AD%A6%E4%B9%A0/%E4%B9%A6%E7%94%9F%C2%B7%E6%B5%A6%E8%AF%AD%E5%A4%A7%E6%A8%A1%E5%9E%8B%E5%AE%9E%E6%88%98/%E7%AC%AC%E5%9B%9B%E8%8A%82%E8%AF%BE%E7%AC%94%E8%AE%B0)

## 动机

将大模型应用到自己所需求领域的数据分为增量预训练和指令跟随能力两种微调方式。

![image-20240114203955330](https://github.com/InternLM/tutorial/assets/54879512/4975e93c-752d-4cdc-a694-bdeed835afae)

1. 增量预训练（SFT）

   base 模型学习新的知识（e.g. 某个垂域的知识）

   数据：文章、书籍、代码

2. 指令跟随微调（IFT）

   模型学会新的对话方式，修改模型的 “说话方式”

   数据：高质量的对话、问答

## 微调方式

微调的流程图：

![image-20240114204040463](https://github.com/InternLM/tutorial/assets/54879512/895913e0-7ae7-4dea-8622-e34d5d9320b0)

### 增量预训练微调

增量训练的数据不需要问答的形式，也就是大模型只需要知道一段话什么时候开始，什么时候结束即可，也可以是陈述句。可以将 System 和 User 留空，只留 Assistant 即可。

![image-20240114205934478](https://github.com/InternLM/tutorial/assets/54879512/9e358985-27ad-4def-bc22-95290b6cc31a)

数据格式：

![image-20240114211041009](https://github.com/InternLM/tutorial/assets/54879512/99c61a2a-dc6f-47aa-b17b-4a376cb363b8)

### 指令跟随微调

构造对话格式数据，对话包含三个角色：

1. System：表示当前对话的上下文信息，可以理解为是一个当前对话环境 Prompt
2. User：表示实际用户会输入的内容，提出一些问题（Input）
3. Assistant：根据 User 的输入，结合当前对话上下文环境（System）做出相应的回答（Output）

三种角色只是方便构造相应格式的数据，以便让 LLM 学习，实际使用并不会感知到这几种角色。

> 假设现在有一问一答的数据，则相应的 Prompt 为：
>
> - System：你是一个地理知识专家
> - User：世界第一高峰是？
> - Assistant：珠穆朗玛峰

![image-20240114211114205](https://github.com/InternLM/tutorial/assets/54879512/2e00a428-cc2c-46cd-a59f-b53b20ab0803)

### 对话模板

为了方便实现上述的流程，通常会设计相应的对话模板，方便在指令微调过程中加入 Prompt 和指令微调的数据。不同的语言模型拥有不同的对话模板。

![image-20240114225553761](https://github.com/InternLM/tutorial/assets/54879512/a7f97b49-c1b1-439a-add8-a040d26f0e8b)

具体形式为：

![image-20240114225738053](https://github.com/InternLM/tutorial/assets/54879512/801dffc9-419c-4998-a411-e0a5170ee9c2)

这些模板都是隐藏在推理代码中，不会展现给用户，用户只需要输入自己的内容即可。

#### Training 过程

指令微调学习主要目的是让模型学会我们指定的回答，也就是希望学会答案（Output）而不是输入的问题（Input），所以计算 Loss 的时候只会对输出进行计算。

![image-20240114225553761](https://github.com/InternLM/tutorial/assets/54879512/ee6752f0-161b-4746-9a5e-b9e208dab8f8)

推理和训练都会加入相应的对话模板：

![image-20240114225738053](https://github.com/InternLM/tutorial/assets/54879512/b792389e-48c0-42d0-b364-19872fc67e8e)

### LoRA & QLoRA

只训练两个低秩矩阵的参数即可，不需要整个网络训练。

![image-20240114234835226](https://github.com/InternLM/tutorial/assets/54879512/a1d2fe29-0d49-484a-b04c-24eed2dbbbce)

## XTuner

- 集成了各种常用的配置文件，能够一键启动微调训练
- 支持各大平台模型
- 具备优化加速

![image-20240114234957566](https://github.com/InternLM/tutorial/assets/54879512/9bd26d63-a560-49f4-9b38-a68a3710dcb1)

### 数据处理

- 数据格式化：具备将原始问答预料转换成对话形式，然后再从对话形式格式化成对话模板的问答模板

- 多数据拼接：能够将多条样本拼接到一条样本上。

  ![image-20240114235240494](https://github.com/InternLM/tutorial/assets/54879512/43f24212-a1ce-4bf3-b7ab-8732cc328d3b)

### 优化加速

- 集成了 Flash Attention
- 集成了 Deepspeed ZeRO

## 总结

利用 XTuner 很方便能够用自己的数据微调大模型，做成一个个人小助手，使得大模型微调能够进入一般用户的视野。

**Comment by abzm11** [comment]
第四课笔记同作业：https://blog.csdn.net/weixin_38043453/article/details/135587465


 user_avator = "/root/personal_assistant/code/InternLM/doc/imgs/user.png"

robot_avator = "/root/personal_assistant/code/InternLM/doc/imgs/robot.png"

ssh -CNg -L 127.0.0.1:6006:0.0.0.0:6006 -o "StrictHostKeyChecking no" root@ssh.intern-ai.org.cn -p 33693

**Comment by lifelinezhang** [comment]
https://zhuanlan.zhihu.com/p/677399381

**Comment by anneeeee** [comment]
https://mp.weixin.qq.com/s/BdP0zBT2dIdel9FpPs2HCg

**Comment by Rogue-Trader-zzy** [comment]
第四次作业 https://www.yuque.com/roguetrader/iu9odz/ef1ui6eig2vsaw0u?singleDoc# 《XTuner大模型单卡低成本微调实践》

**Comment by Rogue-Trader-zzy** [comment]
第四次笔记：https://www.yuque.com/roguetrader/iu9odz/ef1ui6eig2vsaw0u?singleDoc# 《XTuner大模型单卡低成本微调实践》

**Comment by snowhou** [comment]
打卡：https://zhuanlan.zhihu.com/p/679042577

**Comment by Simon-lyz** [comment]
## InternLM Lesson-4
https://g4w90egc30.feishu.cn/docx/K8qodGQ27ojqWrxJxIycyin6nRg?from=from_copylink
